qir/qat/StaticResourcePass/StaticResourcePass.cpp

```diff
@@ -0,0 +1,204 @@
+        {
+            for (auto& instr : block)
+            {
+                for (auto& op : instr.operands())
+                {
+                    auto* a = llvm::dyn_cast<llvm::IntToPtrInst>(op);
+                    auto* b = llvm::dyn_cast<llvm::ConcreteOperator<llvm::Operator, llvm::Instruction::IntToPtr>>(op);
+
+                    auto* nullptr_cast = llvm::dyn_cast<llvm::ConstantPointerNull>(op);
+
+                    if (a || b || nullptr_cast)
+                    {
+
+                        auto pointer_type = llvm::dyn_cast<llvm::PointerType>(op->getType());
+                        if (!pointer_type)
+                        {
+                            continue;
+                        }
+
+                        llvm::Type* element_type = pointer_type->getElementType();
+
+                        if (!element_type->isStructTy())
+                        {
+                            continue;
+                        }
+
+                        auto type_name = static_cast<String>(element_type->getStructName());
+                        if (type_name != "Qubit" && type_name != "Result")
+                        {
+                            continue;
+                        }
+
+                        bool     is_constant_int = nullptr_cast != nullptr;
+                        uint64_t n               = 0;
+
+                        auto user = llvm::dyn_cast<llvm::User>(op);
+
+                        if (user && user->getNumOperands() == 1)
+                        {
+                            auto cst = llvm::dyn_cast<llvm::ConstantInt>(user->getOperand(0));
+
+                            if (cst)
+                            {
+                                is_constant_int = true;
+                                n               = cst->getValue().getZExtValue();
+                            }
+                        }
+
+                        if (is_constant_int)
+                        {
+                            // Updating qubit index
+                            if (type_name == "Qubit")
+                            {
+                                qubits_used.insert(n);
+                                ret.largest_qubit_index = ret.largest_qubit_index < n ? n : ret.largest_qubit_index;
+                            }
+
+                            // Updating result index
+                            if (type_name == "Result")
+                            {
+                                results_used.insert(n);
+                                ret.largest_result_index = ret.largest_result_index < n ? n : ret.largest_result_index;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        ret.usage_qubit_counts  = qubits_used.size();
+        ret.usage_result_counts = results_used.size();
+
+        return ret;
+    }
+
+    void StaticResourcePass::annotateQubits(llvm::Module& module) const
```

Comment:
I think the intention we want to go for is that a transformation pass that changes qubit indexing (either by switching to global indexing from dynamic allocation or as part of a pass that might remap the identifiers) should update the `requiredQubits` annotation. In this case, because this pass does not remap the indices to be sequential without gaps, the required qubits counting can get out of sync with the identifiers used. The pass should remap the qubit identifiers (which may or may not actually result in changes, depending on the initial state) and then update the `requriedQubits` attribute, or be disabled and then leave any existing attributes alone. The combination of remapping plus counting uses into `requiredQubits` should replace the need for a `maxQubitIndex` attribute.

Reply:
I've implemented a first version of remapping. I believe that maxQubitIndex still may be a useful feature, so I would suggest to just make this annotation off by default.